src/filesystem/posix/SDL_sysfsops.c

```diff
@@ -0,0 +1,135 @@
+    if (!dir) {
+        return SDL_SetError("Can't open directory: %s", strerror(errno));
+    }
+
+    struct dirent *ent;
+    while ((retval == 1) && ((ent = readdir(dir)) != NULL))
+    {
+        const char *name = ent->d_name;
+        if ((SDL_strcmp(name, ".") == 0) || (SDL_strcmp(name, "..") == 0)) {
+            continue;
+        }
+        retval = cb(userdata, fs, dirname, name);
+    }
+
+    closedir(dir);
+
+    return retval;
+}
+
+int SDL_SYS_FSremove(SDL_FSops *fs, const char *fullpath)
+{
+    int rc = remove(fullpath);
+    if (rc < 0) {
+        const int origerrno = errno;
+        if (origerrno == ENOENT) {
+            char *parent = SDL_strdup(fullpath);
+            if (!parent) {
+                return -1;
+            }
+            char *ptr = SDL_strrchr(parent, '/');
+            if (ptr) {
+                *ptr = '\0';  // chop off thing we were removing, see if parent is there.
+                struct stat statbuf;
+                rc = stat(parent, &statbuf);
+                if (rc == 0) {
+                    SDL_free(parent);
+                    return 0;  // it's already gone, and parent exists, consider it success.
+                }
+            }
+            SDL_free(parent);
+        }
+        return SDL_SetError("Can't remove path: %s", strerror(origerrno));
+    }
+    return 0;
+}
+
+int SDL_SYS_FSrename(SDL_FSops *fs, const char *oldfullpath, const char *newfullpath)
+{
+    if (rename(oldfullpath, newfullpath) < 0) {
+        return SDL_SetError("Can't remove path: %s", strerror(errno));
+    }
+    return 0;
+}
+
+int SDL_SYS_FSmkdir(SDL_FSops *fs, const char *fullpath)
+{
+    const int rc = mkdir(fullpath, 0770);
+    if (rc < 0) {
+        const int origerrno = errno;
+        if (origerrno == EEXIST) {
+            struct stat statbuf;
+            if ((stat(fullpath, &statbuf) == 0) && (S_ISDIR(statbuf.st_mode))) {
+                return 0;  // it already exists and it's a directory, consider it success.
+            }
+        }
+        return SDL_SetError("Can't create directory: %s", strerror(origerrno));
+    }
+    return 0;
+}
+
+int SDL_SYS_FSstat(SDL_FSops *fs, const char *fullpath, SDL_Stat *st)
+{
+    struct stat statbuf;
+    const int rc = stat(fullpath, &statbuf);
+    if (rc < 0) {
+        return SDL_SetError("Can't stat: %s", strerror(errno));
+    } else if (S_ISREG(statbuf.st_mode)) {
+        st->filetype = SDL_STATPATHTYPE_FILE;
+        st->filesize = (Uint64) statbuf.st_size;
+    } else if (S_ISDIR(statbuf.st_mode)) {
+        st->filetype = SDL_STATPATHTYPE_DIRECTORY;
+        st->filesize = 0;
+    } else {
+        st->filetype = SDL_STATPATHTYPE_OTHER;
+        st->filesize = (Uint64) statbuf.st_size;
+    }
+
+    // SDL file time is seconds since the Unix epoch, so we're already good here.
+    st->modtime = (Uint64) statbuf.st_mtime;
+    st->createtime = (Uint64) statbuf.st_ctime;
+    st->accesstime = (Uint64) statbuf.st_atime;
```

Comment:
Yeah, honestly, I agree. Now that we're not using nanoseconds, I don't see any harm.

(We'll leave file size as unsigned, though. I'm currently of the thinking that if something wants to present a virtual stream of unknown or infinite size, they should mark the thing's filetype as OTHER instead of FILE and give it a size of 0, like it were some sort of Unix device node or whatever.)

Reply:
Changed in f4651ca7c1d5c61b224d4498bccf811d5476c8c3.